### PR TITLE
DWR-1131 Added update method to AssessmentPackageRepository

### PR DESCRIPTION
**Pull request by agorina**

### CODE REVIEW

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java [hunk 1]
```diff
@@ -8,43 +8,21 @@
...
-    private final String naturalId;
-    private final String name;
-    private final int gradeId;
-    private final String label;
-    private final int typeId;
-    private final int subjectId;
-    private final int schoolYear;
-    private final String version;
-    private final Double minScore;
-    private final Double maxScore;
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: Once again, just adjusting to be consistent with the rest of the system.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java [hunk 2]
```diff
@@ -74,7 +52,7 @@ public int getSubjectId() {
         return subjectId;
     }
 
-    public int getsSchoolYear() {
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: Fixed a typo.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java [hunk 3]
```diff
@@ -133,7 +115,41 @@ public Double getCutPoint3() {
...
+            this.typeId = assessmentToCopy.typeId;
+            this.subjectId = assessmentToCopy.subjectId;
+            this.schoolYear = assessmentToCopy.schoolYear;
+            this.label = assessmentToCopy.label;
+            this.version = assessmentToCopy.version;
+            this.minScore = assessmentToCopy.minScore;
+            this.maxScore = assessmentToCopy.maxScore;
+            this.cutPoint1 = assessmentToCopy.cutPoint1;
+            this.cutPoint2 = assessmentToCopy.cutPoint2;
+            this.cutPoint3 = assessmentToCopy.cutPoint3;
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: It is not obvious here why I want to copy without items. Here is an explanation: 
when a package is ingested we will have to Assessments in memory - an existing one and a new one from the package. Then we need to merge them together for an update. While doing it, I am planning on passing a new one to this builder, and then adding items one at a time.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Item.java [hunk 4]
```diff
@@ -143,6 +177,27 @@ public Item build() {
...
+            this.allowCalculator = toCopy.allowCalculator;
+            this.maxPoints = toCopy.maxPoints;
+            this.commonCoreStandards = toCopy.commonCoreStandards;
+            this.otherTargets = toCopy.otherTargets;
+            this.isFieldTest = toCopy.isFieldTest;
+            this.isActive = toCopy.isActive;
+            this.optionsCount = toCopy.optionsCount;
+            this.type = toCopy.type;
+            this.answerKey = toCopy.answerKey;
+            return this;
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Item.java]: Forgot to add tests for this change, working on it.

File: package-processor/src/main/resources/application.sql.yml [hunk 5]
```diff
@@ -63,28 +125,6 @@ sql:
...
-      where a.natural_id in (:asmt_natural_ids)
-
-    commonCoreStandardCreate: >-
-      INSERT INTO item_common_core_standard(item_id, common_core_standard_id)
-          SELECT (SELECT i.id from item i join asmt a on i.asmt_id = a.id where i.natural_id = :item_natural_id and a.natural_id=:asmtId) , :common_core_standard_id
-
-    otherTargetCreate: >-
-      INSERT INTO item_other_target(item_id, target_id)
-          SELECT (SELECT i.id from item i join asmt a on i.asmt_id = a.id  where i.natural_id = :item_natural_id and a.natural_id=:asmtId), :target_id
-
```
Comment (agorina) [inline package-processor/src/main/resources/application.sql.yml]: Did not delete, just moved them around.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentPackageRepository.java [hunk 6]
```diff
@@ -104,12 +106,57 @@ public void createPackage(final Collection<Assessment> assessments, final long i
...
+        if (!assessmentNaturalIds.isEmpty()) jdbcTemplate.update(sqlUpdateDifficultyCode, new MapSqlParameterSource("asmt_natural_ids", assessmentNaturalIds));
+    }
+
     @Transactional(readOnly = true)
     @Override
     public Assessment findOneByNaturalId(final String id) {
 
         final List<Assessment.Builder> assessmentList = jdbcTemplate.query(sqlFindOneByNaturalId, new MapSqlParameterSource("natural_id", id), new AssessmentRowMapper());
-        if (assessmentList.size() == 1) {
+        if (!assessmentList.isEmpty()) {
```
Comment (agorina) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentPackageRepository.java]: @jtreuting , addressing your comment from the last review.

Comment (jtreuting) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentPackageRepository.java]: that will make it 100 times faster now ;)

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/model/ItemTest.java [hunk 7]
```diff
@@ -2,108 +2,144 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
 /**
  * Unit test for Item class
  */
 public class ItemTest {
 
     @Test(expected = NullPointerException.class)
     public void itShouldNotAllowNullDokParameter() {
-        final Item.Builder builder = Item.builder()
```
Comment (agorina) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/model/ItemTest.java]: My OCD was the source of these changes.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentPackageRepository.java (hunk 6 above)
Comment (agorina) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentPackageRepository.java]: It better be!  :)

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultItemParserIT.java [hunk 8]
```diff
@@ -54,8 +54,8 @@ public void itShouldCatchAllErrorsAndReturnNoObjectsWhenParsingBadData() throws
                 "{\"elementName\":\"[(SBAC)SBAC-ICA-FIXED-G11M-Perf-TeenDrivingRestricitons-MATH-11-Winter-2016-2017], item id [200-21564], AllowCalculator\",\"value\":\"something\",\"error\":\"invalid value [something]\"}," +
                 "{\"elementName\":\"[(SBAC)SBAC-ICA-FIXED-G11M-Winter-2016-2017], item id [200-12585], MathematicalPractice\",\"value\":\"something\",\"error\":\"For input string: \\\"something\\\"\"}," +
                 "{\"elementName\":\"[(SBAC)SBAC-ICA-FIXED-G11M-Winter-2016-2017], item id [200-501], MaxPoints\",\"value\":\"\",\"error\":\"value may not be blank\"}," +
-                "{\"elementName\":\"[(SBAC)SBAC-ICA-FIXED-G11M-Winter-2016-2017], item id [200-501], DOK\",\"value\":\"662\",\"error\":\"invalid value\"}," +
                 "{\"elementName\":\"[(SBAC)SBAC-ICA-FIXED-G11M-Winter-2016-2017], item id [200-501], CommonCore\",\"value\":\"ABCD;ABCF\",\"error\":\"unable to find all common core standards for an item and subject id [1]\"}," +
+                "{\"elementName\":\"[(SBAC)SBAC-ICA-FIXED-G11M-Winter-2016-2017], item id [200-501], DOK\",\"value\":\"662\",\"error\":\"invalid value\"}," +
```
Comment (agorina) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultItemParserIT.java]: Nothing changed, just error msg appear in a diff order.

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentPackageRepositoryIT.java [hunk 9]
```diff
@@ -142,10 +145,8 @@ public void itShouldFindOneByNaturalId() {
 
         final List<Item> items = assessment.getItems();
         assertThat(items.size()).isEqualTo(4);
-        Item item990 = null;
-        for (final Item item : items) {
-            if (item.getId() == -990) item990 = item;
-        }
+
+        final Item item990 = items.stream().filter(i -> i.getId() == -990).collect(toList()).get(0);
```
Comment (malaffoon) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentPackageRepositoryIT.java]: i think `.findFirst().get()` is a better pattern than `.collect(toList()).get(0)`

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java (hunk 3 above)
Comment (jtreuting) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: maybe there should be a comment in the code explaining that as I'm sure in the future we will forget

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Item.java [hunk 10]
```diff
@@ -153,9 +208,7 @@ public Integer getId() {
         }
 
         public Builder commonCoreStandards(final Collection<Integer> commonCoreStandards) {
-            if (commonCoreStandards != null) {
-                this.commonCoreStandards.addAll(commonCoreStandards);
-            }
+            this.commonCoreStandards = commonCoreStandards == null ? newHashSet() : newHashSet(commonCoreStandards);
```
Comment (jtreuting) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Item.java]: am i reading this wrong.  above on the getCOmmonCoreStandards() it is using a newArrayList and here it is using a newHashSet.  Shouldn't those be the same?  Does the order matter on them?

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentPackageRepositoryIT.java (hunk 9 above)
Comment (agorina) [reply package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentPackageRepositoryIT.java]: Will change.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java (hunk 3 above)
Comment (agorina) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: I thought that having an explicit name  "copyWithoutItems" is self-explanatory of what it does?

Comment (agorina) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: Good point, I missed that. 
Will change to the Set everywhere. 
The order does not matter, but the goal is to filter the duplicates.

Comment (jtreuting) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: goods point :) . i didn't even notice the name and was more focused on the fact that you thought you needed to explain why it wasn't doing that in a comment in GitHub.  but in the code you will be able to tell that it is using that method and that it won't bring items along

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Item.java (hunk 10 above)
Comment (agorina) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Item.java]: Good point, I missed that. 
Will change to the Set everywhere. 
The order does not matter, but the goal is to filter the duplicates.

**Pull request merged**